### PR TITLE
Implement resample2d

### PR DESCRIPTION
**Pull request by BruceDai**
@fdwr Thanks much for explanations of resample2d by `nearest-neighbor` and `linear` modes on #18 and https://github.com/webmachinelearning/webnn/issues/358, this PR is to sample pixel centers to align with WebNN-Polyfill and Chromium WebNN implementation.

@fdwr @huningxin @lisa0314 PTAL, thanks.

### CODE REVIEW

File: src/resample2d.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ * @return {Tensor}
+ */
+export function resample2d(
+    input,
+    {
+      mode = 'nearest-neighbor',
+      scales = [1.0, 1.0],
+      sizes,
+      axes = [2, 3],
+    } = {}) {
```
Comment (fdwr) [inline src/resample2d.js:114]: (minor) It's kinda hard to follow these `} = {}) {`. Line break somewhere in there to aid readability?

File: src/resample2d.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  if (axes[0] === 0) {
+    // hwnc -> nhwc
+    input = transpose(input, {permutation: [2, 0, 1, 3]});
```
Comment (fdwr) [inline src/resample2d.js:117]: It would be more robust to verify that `axes[0] == 0 && axes[1] == 1`, rather than just `axes[0]`.

I'd also add an `else` for `axes[0] == 1 && axes[1] == 2` for clarity, even if it is just a comment that says "NHWC -> NHWC Do nothing". That way it's apparent to readers that you didn't leave it out.

Comment (fdwr) [reply src/resample2d.js:117]: Btw, I had this clever idea that you could just say a single line...

`input = transpose(permutation: [(axes[0] + 2) & 3, (axes[1] + 2) & 3, axes[0], axes[1]]);`

...instead of explicit switch case statements for each transpose, and it works nicely and generically for the first transpose (it assigns the last 2 dimensions H and W to read from whatever axes were given, and it assigns the batch and channel to the other two axes). All the standard axes work:

`axes = [0,1] -> permute:[2, 3, 0, 1]`
`axes = [1,2] -> permute:[3, 0, 1, 2]`
`axes = [2,3] -> permute:[0, 1, 2, 3]`

And it works for the *mirrors* of the standard ones:

`axes = [1,0] -> permute:[3, 2, 1, 0]`
`axes = [2,1] -> permute:[0, 3, 2, 1]`
`axes = [3,2] -> permute:[1, 0, 3, 2]`

And even works for HNCW (which isn't valid per the spec anyway):

`axes = [0,3] -> permute:[2, 1, 0, 3]`

But it's not clear how to *untranspose* that (the 2nd transpose). So, I'm not proposing that idea...

File: test/resample2d_test.js [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+          mode: 'nearest-neighbor',
+          sizes: [4, 6],
+          axes: [2, 3],
+        },
+        {
+          shape: [1, 1, 4, 6],
+          data: [
+            1, 1, 1, 2, 2, 2, 1, 1, 1, 2, 2, 2,
+            3, 3, 3, 4, 4, 4, 3, 3, 3, 4, 4, 4,
+          ],
```
Comment (fdwr) [inline test/resample2d_test.js:266]: The output would be clearer if wrapped consistently with the shape. e.g.

          data: [
            1, 1, 1, 2, 2, 2,
            1, 1, 1, 2, 2, 2,
            3, 3, 3, 4, 4, 4,
            3, 3, 3, 4, 4, 4,
          ],

Like you nicely did with the other cases above. e.g.

          shape: [1, 1, 4, 4],
          data: [
            1.0, 1.25, 1.75, 2.0,
            1.5, 1.75, 2.25, 2.5,
            2.5, 2.75, 3.25, 3.5,
            3.0, 3.25, 3.75, 4.0,
          ],

File: src/resample2d.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+ * @param {Tensor} src
+ * @param {Tensor} dst
+ * @param {Number} dstHeight
+ * @param {Number} dstWidth
+ * @param {Number} scaleHeight
+ * @param {Number} scaleWidth
+ * @return {Tensor}
+ */
+function nearestNeighbor(src, dst, dstHeight, dstWidth, scaleHeight, scaleWidth) {
+  const [srcBatches, srcHeight, srcWidth, srcChannels] = src.shape;
```
[the remark this answers is not in the document]
Comment (BruceDai) [reply src/resample2d.js]: Good catch! Now I let them be the same of NCHW, thanks.

File: src/resample2d.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  for (let ob = 0; ob < srcBatches; ob++) {
+    for (let oc = 0; oc < srcChannels; oc++) {
+      for (let oh = 0; oh < dstHeight; oh++) {
+        let ih = (oh + 0.5) / scaleHeight - 0.5;
+        if (ih < 0) {
+          ih = 0;
+        } else if (ih > srcHeight - 1) {
+          ih = srcHeight - 1;
+        }
```
[the remark this answers is not in the document]
Comment (BruceDai) [reply src/resample2d.js]: Updated, thanks.

File: src/resample2d.js (hunk 1 above)
Comment (BruceDai) [reply src/resample2d.js]: Updated, thanks.

File: test/resample2d_test.js (hunk 3 above)
Comment (BruceDai) [reply test/resample2d_test.js:308]: Updated test data, thanks.

File: src/resample2d.js (hunk 2 above)
Comment (BruceDai) [reply src/resample2d.js]: > input = transpose(permutation: [(axes[0] + 2) & 3, (axes[1] + 2) & 3, axes[0], axes[1]]);'

Thanks, this code is fairly concise. 
For more easily understand, I modified previous switch case statements following your first comment.

**Comment by BruceDai** [comment]
@fdwr Thanks for your reviewing! I updated commit to address your comments, please take another look, thanks.

**Review by fdwr — APPROVED** [review]
:)

**Review by huningxin — APPROVED** [review]
